Adjust fee currency blacklisting

piersy opened the pull request.

piersy committed:
Add rpc call to enable/disable the fee currency blocklist

ezdac committed:
Add setter for currency blocklist filter status

mcortesi commented on miner/hash_blocklist.go:
```diff
@@ -0,0 +1,20 @@
+var TxHashEvictionTimeoutSeconds uint64 = 24 * hours
+
+func TxHashReachedEvictionTimeout(h *types.Header, latest *types.Header) bool {
+	return h.Time+TxHashEvictionTimeoutSeconds < latest.Time
```
if i understand correctly you're storing the whole header for each TX, which makes the blocklist rather big in memeroy, i'd store just the eviction date.. or something..

keep it as simple/small as possible

palango commented on eth/api_admin.go:
```diff
@@ -141,3 +141,8 @@ func (api *AdminAPI) ImportChain(file string) (bool, error) {
 }
+
+// ImportChain imports a blockchain from a local file.
+func (api *AdminAPI) SetfeeCurrencyBlocklistStatus(enabled bool) {
```
```suggestion
func (api *AdminAPI) SetFeeCurrencyBlocklistStatus(enabled bool) {
```

ezdac committed:
Allow to set filter status on address blocklist

ezdac committed:
Fix camelcase in admin api method

piersy committed:
Add tx ringbuffer

ezdac committed:
Fix invalidating currency gas-pool when blocking disabled

palango commented on miner/currency_blocklist.go:
```diff
@@ -18,21 +19,41 @@ type AddressBlocklist struct {
 	headerEvictionTimeoutSeconds uint64
 	oldestHeader                 *types.Header
+
+	enactFilterEnabled *atomic.Int32
```
Why not use [`atomic.Bool`](https://pkg.go.dev/sync/atomic#Bool)?

ezdac replied:
That's a very good question... somehow I thought it doesn't exist

ezdac committed:
Add RW mutex to tx-ringbuffer

ezdac committed:
Fix adding tx to ringbuffer multiple times

seolaoh commented on miner/worker.go:
```diff
@@ -838,8 +840,10 @@ func (miner *Miner) blockFeeCurrency(env *environment, feeCurrency common.Addres
 	// transactions.
-	pool := env.multiGasPool.PoolFor(&feeCurrency)
-	pool.SetGas(0)
+	if env.blockingAllowed {
```
Why don't we handle this where this function is called? like [here](https://github.com/celo-org/op-geth/blob/08aa267aac4f5ade9c275a64229bac7ea34c25b7/miner/worker.go#L429)?

piersy replied:
Setting the pool gas to zero is part of the blocking action, so keeping it here in this function keeps relevant actions in one place.

Why would you want to move that out?

seolaoh replied:
I meant something like this,
```go
if env.blockingAllowed {
  miner.blockFeeCurrency(env, *tx.FeeCurrency(), err)
}
```
just wondering if there is any reason not to trigger the whole function if blocking allowed

ezdac replied:
We decided that we still want to *notice/register* that something went wrong in the blocklist, but we want to enable/disable *applying* that blocklist for payload building.

Also, we still want to block on the transaction level, even if we don't apply blocking the whole fee-currency.
For that we need to add the tx-hash to the ringbuffer some lines below.

piersy replied:
Ok got you, it's because we do still execute some blocking related actions even when we've prevented blocking.

To clarify, the approach is to maintain the blocklist as it normally would be but not to enforce the block list if blocking is not allowed.

piersy committed:
Connect transaction blocking

piersy committed:
Add admin methods

ezdac committed:
Add unblock-fee currency method to miner

piersy committed:
Add unblock transaction

ezdac committed:
Remove nil check in fee-currency-blocklist setter

We want to be verbose when the API handler fails, since the blocklist is
assumed to not be nil

ezdac committed:
Use atomic.Bool in address-blocklist

piersy committed:
Fix ringbuffer remove locking

piersy committed:
Return bool to indicate whether removal actions had any effect

mcortesi commented on miner/tx_ringbuffer.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	// Clean up old transaction hash from map
+	oldHash := b.ring[b.index]
```
wonder what happens when the ring is not yet full... this returns zero and delete(zero) doesn't break?

piersy replied:
It will be the empty hash, and delete the empty hash which will be a no-op.

mcortesi commented on miner/worker.go:
```diff
@@ -345,6 +346,7 @@ func (miner *Miner) prepareWork(genParams *generateParams, witness bool) (*envir
 		common.CurrencyAllowlist(env.feeCurrencyContext.ExchangeRates),
 		header,
 	)
+	env.blockingAllowed = miner.feeCurrencyBlocklist.GetEnableFilterStatus()
```
is this done for every block?

piersy replied:
Yep, env is constructed per block at line 331

ezdac replied:
Yes, every time a new payload is requested to be built by the engine-api a new `environment` is created

mcortesi commented on miner/worker.go:
```diff
@@ -596,6 +599,12 @@ func (miner *Miner) commitTransactions(env *environment, plainTxs, blobTxs *tran
 			}
 		}
+		if miner.blockedTxRingbuffer.Has(ltx.Hash) {
+			log.Trace("Skipping tx execution, tx in blockedTxRingbuffer", "hash", ltx.Hash)
```
trace? maybe some level we had for skipping a tx due to fee currency lock

ezdac replied:
This would be logged on every block build attempt, so every 1 second

piersy replied:
This is the same level we had for skipping a tx due to the fee currency lock, see line 597

mcortesi commented on miner/worker.go:
```diff
@@ -834,14 +843,19 @@ func (miner *Miner) validateParams(genParams *generateParams) (time.Duration, er
 	// so that they are not allowlisted in the following blocks
 	// (only locally in the txpool, not consensus-critical)
-	miner.feeCurrencyBlocklist.Add(feeCurrency, *env.header)
+	miner.feeCurrencyBlocklist.Add(*tx.FeeCurrency(), *env.header)
```
we're sure that tx.FeeCurrency is never nil here?

piersy replied:
Previously this function was being called with the de-referenced fee currency [here](https://github.com/celo-org/op-geth/pull/345/files#diff-689874b30f6f905ce6c47cdefeddcf052b467a7936a981f4b6cf38939e10d924L427) (line 427, _github seems to have an issue correctly scrolling to that point_)  so I'm as sure as we previously were that fee currency is not nil.

seolaoh commented on miner/miner.go:
```diff
@@ -231,6 +233,17 @@ func (miner *Miner) getPending() *newPayloadResult {
 
+func (miner *Miner) SetFeeCurrencyBlocklistStatus(enabled bool) {
+	miner.feeCurrencyBlocklist.SetEnableFilterStatus(enabled)
+}
```
```suggestion
}

```

seolaoh commented on miner/miner.go:
```diff
@@ -231,6 +233,17 @@ func (miner *Miner) getPending() *newPayloadResult {
+func (miner *Miner) SetFeeCurrencyBlocklistStatus(enabled bool) {
+	miner.feeCurrencyBlocklist.SetEnableFilterStatus(enabled)
+}
+func (miner *Miner) UnblockFeeCurrency(address common.Address) {
```
How about returning the returned value of `miner.feeCurrencyBlocklist.Remove(address)` and consequently returning the request status in `func (api *AdminAPI) UnblockFeeCurrency(address common.Address)`? To know if the request was successful or not when we call the admin API.

ezdac replied:
Good idea! I think we have implemented this already - @piersy am I correct and you didn't push the changes yet?

piersy replied:
Yep, pushed now, forgot to push yesterday

seolaoh commented on eth/api_admin.go:
```diff
@@ -141,3 +142,15 @@ func (api *AdminAPI) ImportChain(file string) (bool, error) {
+	api.eth.Miner().UnblockTransaction(hash)
+}
+
+func (api *AdminAPI) UnblockFeeCurrency(address common.Address) {
```
Would it be helpful to have getter API for blocked fee currency list or tx list?

ezdac replied:
Why not. I think the fee-currency is not strictly necessary since we have the "block" and "evict" log messages.
The transactions would be helpful, because this is only logged with `trace` level.

piersy replied:
Yeah I think so. TBH I want to go over this whole blocking feature again and ensure the design is clean. Yesterday we were somewhat in a hurry so didn't have the time to consider all the niceties that you normally would.

seolaoh commented on miner/tx_ringbuffer.go:
```diff
@@ -0,0 +1,65 @@
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	delete(b.m, tx)
```
Don't we need to handle `b.index` and `b.ring` also?

piersy replied:
I don't think so.

The idea here is that the ring buffer is large enough that transactions added don't get overwritten for a pretty long time, we made it 100000. So far on the L2 we've had about 750 txs paid with USD, so 100000 should allow txs to exist in the buffer for a really long time till they get overwritten. So although we could update the index and  remove the tx from the queue, I don't think its really necessary. The purpose of removal is to stop that transaction from being blocked which is done, it doesn't matter that the hash stays in the ring's slice.

Let me know if you think there's a problem with this approach?

seolaoh replied:
Oh I see your point, I thought we wanna fully utilize the 100000 capacity. In that case, if we remove specific txs, the capacity would be less than 100000 until the index reaches the removed entry. But I agree that it's not a required feature.

piersy committed:
Update miner/miner.go

Co-authored-by: Seola Oh <osa8361@gmail.com>

ezdac committed:
Add metrics-counter to blocklists

ezdac commented:
I pulled out the fixes for deactivating the blocklist on l1 derivation / pending building to #363.
This PR is separate (without the changes here), but I have a local copy of the branch that keeps the changes in this branch here into account.

seolaoh commented on miner/currency_blocklist.go:
```diff
@@ -18,21 +19,40 @@ type AddressBlocklist struct {
+	return b.enactFilterEnabled.Load()
+}
+
+// This only activates / deactivates wether the
```
```suggestion
// SetEnableFilterStatus only activates / deactivates whether the
```

seolaoh commented on miner/worker.go:
```diff
@@ -418,13 +424,14 @@ func (miner *Miner) commitTransaction(env *environment, tx *types.Transaction) e
 	receipt, err := miner.applyTransaction(env, tx)
 	if err != nil {
 		if errors.Is(err, contracts.ErrFeeCurrencyEVMCall) {
+
```
```suggestion
```

seolaoh reviewed (approved):
LGTM, just some nits

piersy commented:
> LGTM, just some nits

Hey @seolaoh, thanks we actually decided to split this up, so there will be some smaller PRs arriving soon, first one here - https://github.com/celo-org/op-geth/pull/363

piersy commented:
Closing in favour of 3 more targeted PRs
